src/engine/specification.md

```diff
@@ -61,11 +61,12 @@ The versioning of the Engine API is defined as follows:
 
 ## Message ordering
 
-Consensus Layer client software **MUST** utilize JSON-RPC request IDs that are strictly
-increasing.
+Consensus Layer client software **MUST** utilize JSON-RPC request IDs that are strictly increasing.
+Request IDs assigned to `engine_forkchoiceUpdated` method calls **MUST** respect the order of
```

Comment:
I worry this could lead to accidental deadlock scenarios. 

For example:
* CL1 uses JSON-RPC id's up to 100
* CL1 is sawpped for CL2. CL2 starts at ID 0
* EL won't listen to CL2 forkchoice updates until it exceeds 100

Maybe EL resets their local ID max per-session?

Or maybe we find a better method. I think the session reset works, but overriding this mechanism seems a bit brittle

Reply:
I concur. We don't have sessions yet, probably we will have them later on --  depends on the authentication machinery we end up with for Engine API.

Ideally, we tie the order to intrinsic property of communicated fork choice state. For example, it could be a slot number which head block does belong to. But this would not always work, as switching the head to a block from the previous slot is pretty legal. There is no intrinsic property that we can rely on, while the issue you have mentioned seems to exist with any artificial ordering scheme.

So, we seem to need sessions to be able to handle message ordering properly.

Reply:
I'm okay with reset ID

`RPC_ORDER_RESET_ID = 0` 
If `msg.id == RPC_RESET_ID`, rest on EL side.

TODO: Specify what happens to messages that happened in the prior ID session. Are they forgotten/dropped?

Funny case to think through -- what happens if CL *always* uses 0

Reply:
> TODO: Specify what happens to messages that happened in the prior ID session. Are they forgotten/dropped?

I guess, if such messages were received and cached somewhere, they will be processed before processing the reset message.

> Funny case to think through -- what happens if CL always uses 0

CL client doing so doesn't follow the spec hence is broken

Reply:
Ok well im still trying to get a verified message signature